contrib/pyudmi/src/udmi/envelope.py

```diff
@@ -0,0 +1,65 @@
+"""
+{
+  "title": "Message envelope schema",
+  "additionalProperties": true,
+  "properties": {
+    "deviceId": {
+      "type": "string",
+      "pattern": "^[A-Z]{2,6}-[0-9]{1,6}$"
+    },
+    "deviceNumId": {
+      "type": "string",
+      "pattern": "^[0-9]+$"
+    },
+    "deviceRegistryId": {
+      "type": "string",
+      "pattern": "^[a-zA-Z][-a-zA-Z0-9._+~%]*[a-zA-Z0-9]$"
+    },
+    "projectId": {
+      "type": "string",
+      "pattern": "^([.a-z]+:)?[a-z][-a-z0-9]*[a-z0-9]$"
+    },
+    "subFolder": {
+      "enum": [
+        "config",
+        "discover",
+        "system",
+        "metadata",
+        "pointset",
+        "state"
+      ]
+    }
+  },
+  "required": [
+    "projectId",
+    "deviceRegistryId",
+    "deviceNumId",
+    "deviceId",
+    "subFolder"
+  ]
+}
+
+"""
+
+from .base import UDMIBase, DEFAULT_UDMI_VERSION
+
+
+class Envelope(UDMIBase):
+
+    schema = "envelope.json"
+    __slots__ = [
```

Comment:
what's the effective impact of the __slots__? From reading online, it seems to be a memory optimization, useful for when you're creating "thousands or millions" of objects... was there something else this was for?  It seems like a lot of extra overhead (the __slots__) for something that is likely to have much impact (when are there going to be so many objects?)

Reply:
yes - slots are there mostly as an optimisation. I thought it might be useful if working with large collections of pointsets. But then i'm also reusing them for serialisation. This would be better done directly from the schemas.

Reply:
yes - the slots are mostly an optimisation, which i though might be useful if working with large collections of pointsets. But I have also reused them in the serialisation to json, this could/should be done by referring directly to the schemas instead, then the slots could be removed.

Reply:
I have  tried the json codegen libs but they might be useful (they've changed the way these comments work and I can't work out how to inline comment below)